### PR TITLE
markdown: <TOC/> component, fix #1275 [WIP]

### LINKED ISSUE #1275: TOC links to header with badge fail
- [x] I confirm that this is a issue rather than a question.

## Bug report
When a header includes a badge, a TOC will report an incorrect scroll link for that header.

#### Version
0.14.8
#### Steps to reproduce
1. go to https://v0.vuepress.vuejs.org/guide/markdown.html#table-of-contents
2. click "Import Code Snippets"

#### What is expected?
It should scroll to the header link
#### What is actually happening?
nothing happens
#### Other relevant information

- Your OS: Ubuntu 18.10
- Node.js version: 10.15.1
- Browser version: chrome 72
- Is this a global or local install? local
- Which package manager did you use for the install? yarn

### PR DESCRIPTION
**Pull request by Shigma**
**Summary**

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Docs
- [ ] Build-related changes
- [ ] Other, please describe:

If changing the UI of default theme, please provide the **before/after** screenshot:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)

You have tested in the following browsers: (Providing a detailed version will be better.)

- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge
- [ ] IE

If adding a **new feature**, the PR's description includes:

- [ ] A convincing reason for adding this feature
- [x] Related documents have been updated
- [ ] Related tests have been updated

To avoid wasting your time, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

- Be compatible with old `[[TOC]]` syntax while enable new syntax `<TOC/>`. You can configurate TOC in `markdown.toc` or by directly passing props to `<TOC/>` component.
- BREAKING: `slugify`, `format` and `forceFullToc` are not supported at present.
- Documentation has been updated as well.

### CODE REVIEW

**Comment by Shigma** [comment]
@ulivz could you have a review on this? The original issue is marked as contribution welcome.

File: packages/@vuepress/markdown/components/TOC.vue [hunk 1]
```diff
@@ -0,0 +1,65 @@
+<template>
+  <div>
+    <p class="header" v-html="containerHeaderHtml" />
+    <HeaderList :items="groupedHeaders" :list-type="listTypes" />
+    <p class="footer" v-html="containerFooterHtml" />
```
Comment (ulivz) [inline packages/@vuepress/markdown/components/TOC.vue]: I guess we can avoid using the v-html here.

File: packages/docs/docs/config/README.md [hunk 2]
```diff
@@ -222,7 +222,12 @@ The key and value pair will be added to `<a>` tags that points to an external li
 - Type: `Object`
 - Default: `{ includeLevel: [2, 3] }`
 
-Options for [markdown-it-table-of-contents](https://github.com/Oktavilla/markdown-it-table-of-contents). (Note: prefer `markdown.slugify` if you want to customize header ids.)
+This attribute will control the behaviour of `[[TOC]]`. We design this interface as close as possible to the options of [markdown-it-table-of-contents](https://github.com/Oktavilla/markdown-it-table-of-contents). However, there are some little differences:
+
+1. `slugify`, `format` and `forceFullToc` are not supported at present.
```
Comment (ulivz) [inline packages/docs/docs/config/README.md]: We don't need to support `slugify` individually for this TOC component or the link would go the 404.

Unsupported options aren't recommended in the docs, it should be mentioned as a breaking change in the changelog.

**Comment by Shigma** [comment]
> Thanks for the great work, but I don't want to introduce so much code to fix this issue.

😢 I thought those addition could be compromised by removing a dependency.

**Comment by ulivz** [comment]
Another solution is that you can transform your code changes into an internal plugin, for now the code are too dispersed.

**Comment by Shigma** [comment]
@ulivz I added `TOC` into *built-in-components*. What badge do you plan for the `TOC` component?

**Comment by Shigma** [comment]
## Breaking Changes and Migrations

### slugify
- use `markdown.slugify` instead.

### format
- currently not supported by `$page.headers`.
- also we cannot transfer function from server to client.
- *future solutions*: `markdown.format`: function for formatting headings.

### forceFullToc
- currently not supported by `$page.headers`.

**Comment by lunaceee** [comment]
I'm not sure if this is relevant but before I create a new issue I'd like you guys to verify my question first:

I'd like to add a `toc` to the site level, instead of adding it to each markdown file. 

I tried adding 
```<TOC :list-type="['ol', 'ul']"><p slot="header"><strong>Custom Table of Contents</strong></p</TOC>```
to the `Layout.vue` file, got the following error message:
```<TOC> - did you register the component correctly? For recursive components, make sure to provide the "name" option.```

Any suggestions are highly appreciated!

**Comment by Shigma** [comment]
@lunaceee I think it's not our problem.

https://www.meteorlxy.cn/posts/2018/11/06/koa-core-read-source-code-part-3.html

You see, this is a website powered by VuePress, which indeed has a TOC on the site level, and it works well.

**Comment by lunaceee** [comment]
Thanks @Shigma, this is really helpful!

**Pull request closed without merging**